File: packages/cli/src/commands/evaluate.ts
/**
 * `simforge evaluate <trace> [--filter critical|negative-control|all]`.
 *
 * The reject filters, applied to a trace. `--filter negative-control` is not a
 * different set of filters — it is the same set with `trivially_safe` demoted
 * from a rejection to a tag, because the taxonomy deliberately contains
 * scenarios whose whole point is that nothing happens.
 */

import { readFile } from 'node:fs/promises';

import { intentRubricSchema, type EvaluateFilters, type IntentEvaluation } from '@simforge-oss/engine';
import { readTraceHandle, writeJsonFile } from '@simforge-oss/compiler/node';

import { CliError, EXIT } from '../errors.js';
import { emit, emitLines, fixed, pad } from '../output.js';
import { metricsSummary } from './simulate.js';

export type EvaluateFilterMode = 'critical' | 'negative-control' | 'all';

export interface EvaluateOptions {
  readonly file: string;
  readonly filter: EvaluateFilterMode;
  readonly trivialTtcS?: number | undefined;
  readonly rejectCollisions: boolean;
  /** Optional canonical IntentRubric JSON evaluated alongside criticality. */
  readonly rubric?: string | undefined;
  /** Optional bounded packet for a context-blind Codex reviewer. */
  readonly blindReviewOut?: string | undefined;
  readonly pretty: boolean;
}

/**
 * Intent evidence may explain an intentionally stationary/no-conflict episode,
 * but it never suppresses hard generic safety or execution failures.
 */
export function combinedEvaluationVerdict(
  generic: { verdict: 'accept' | 'reject'; findings: ReadonlyArray<{ code: string }> },
  intent: IntentEvaluation | null,
): 'accept' | 'reject' {
  if (intent === null) return generic.verdict;
  if (intent.verdict === 'reject') return 'reject';
  const explainedByIntent = new Set(['no_interaction', 'trivially_safe', 'out_of_window']);
  return generic.findings.some((finding) => !explainedByIntent.has(finding.code)) ? 'reject' : 'accept';
}

async function readIntentRubric(file: string) {
  let raw: unknown;
  try {
    raw = JSON.parse(await readFile(file, 'utf8'));
  } catch (error) {
    throw new CliError('invalid_json', `cannot read intent rubric ${file}: ${error instanceof Error ? error.message : String(error)}`, { path: file });
  }
  const parsed = intentRubricSchema.safeParse(raw);
  if (!parsed.success) {
    throw new CliError('bad_value', 'the intent rubric is invalid', {
      path: file,
      detail: { issues: parsed.error.issues.map((issue) => ({ path: issue.path.join('.'), reason: issue.message })) },
      exitCode: EXIT.validationFindings,
    });
  }
  return parsed.data;
}

export function filtersFor(
  mode: EvaluateFilterMode,
  extra: { trivialTtcS?: number | undefined; rejectCollisions?: boolean } = {},
): EvaluateFilters {
  return {
    ...(mode === 'negative-control' ? { negativeControl: true } : {}),
    ...(extra.trivialTtcS === undefined ? {} : { trivialTtcS: extra.trivialTtcS }),
    ...(extra.rejectCollisions ? { rejectCollisions: true } : {}),
  };
}

/** `critical` / `trivially-safe` / `infeasible` — the batch's headline buckets. */
export function criticalityBand(
  verdict: 'accept' | 'reject',
  findings: ReadonlyArray<{ code: string }>,
): 'critical' | 'trivially-safe' | 'no-interaction' | 'unavoidable' | 'out-of-window' | 'never-fired' {
  if (verdict === 'accept') return 'critical';
  const codes = new Set(findings.map((f) => f.code));
  if (codes.has('no_interaction')) return 'no-interaction';
  if (codes.has('trivially_safe')) return 'trivially-safe';
  if (codes.has('physically_unavoidable')) return 'unavoidable';
  if (codes.has('never_fired')) return 'never-fired';
  if (codes.has('out_of_window')) return 'out-of-window';
  return 'trivially-safe';
}

export async function evaluate(options: EvaluateOptions): Promise<number> {
  if (!['critical', 'negative-control', 'all'].includes(options.filter)) {
    throw new CliError('bad_value', `--filter must be critical | negative-control | all`, {
      path: '--filter',
    });
  }
  const handle = await readTraceHandle(options.file);
  const trace = handle.toTrace();
  const evaluation = handle.evaluate(
    filtersFor(options.filter, {
      trivialTtcS: options.trivialTtcS,
      rejectCollisions: options.rejectCollisions,
    }),
  );
  const band = criticalityBand(evaluation.verdict, evaluation.findings);
  let intentEvaluation: IntentEvaluation | null = null;
  if (options.rubric) {
    const rubric = await readIntentRubric(options.rubric);
    intentEvaluation = handle.evaluateIntentRubric(rubric);
    if (options.blindReviewOut) {
      await writeJsonFile(options.blindReviewOut, handle.blindReviewPacket(rubric));
    }
  } else if (options.blindReviewOut) {
    throw new CliError('missing_argument', '--blind-review-out requires --rubric', { path: '--rubric' });
  }
  const combinedVerdict = combinedEvaluationVerdict(evaluation, intentEvaluation);

  const payload = {
    file: options.file,
    mapId: trace.header.mapId,
    metricSubject: trace.header.metricSubject,
    filter: options.filter,
    verdict: evaluation.verdict,
    combinedVerdict,
    band,
    tags: evaluation.tags,
    findings: evaluation.findings,
    summary: evaluation.summary,
    metrics: metricsSummary(trace),
    ...(options.rubric ? { intentRubric: options.rubric, intentEvaluation } : {}),
    ...(options.blindReviewOut ? { blindReviewPacket: options.blindReviewOut } : {}),
  };

  if (!options.pretty) {
    emit(payload, options);
  } else {
    const lines = [
      `${options.file}: ${combinedVerdict.toUpperCase()} (${band}; generic ${evaluation.verdict})`,
      `minTTC ${fixed(evaluation.summary.minTTC)} s at t=${fixed(evaluation.summary.minTTCt)} s · requiredDecelMax ${fixed(
        evaluation.summary.requiredDecelMax,
      )} m/s² · collisions ${evaluation.summary.collisions} · neverFired ${evaluation.summary.neverFired}`,
    ];
    if (evaluation.tags.length > 0) lines.push(`tags: ${evaluation.tags.join(', ')}`);
    if (evaluation.findings.length > 0) {
      lines.push('', 'findings:');
      for (const f of evaluation.findings) lines.push(`  ${pad(f.code, 26)}${f.reason}`);
    }
    if (intentEvaluation) {
      lines.push('', `intent: ${intentEvaluation.verdict.toUpperCase()} · ${intentEvaluation.counts.pass} pass · ${intentEvaluation.counts.fail} fail · ${intentEvaluation.counts.unchecked} unchecked · ${intentEvaluation.counts.unsupported} unsupported`);
      for (const criterion of intentEvaluation.criteria) lines.push(`  ${pad(criterion.status, 12)}${criterion.id}: ${criterion.reason}`);
    }
    emitLines(lines);
  }

  // A rejection is a *finding*, not a command failure: the caller asked whether
  // this instance passes, and it got a definite answer.
  return combinedVerdict === 'accept' ? EXIT.ok : EXIT.validationFindings;
}

File: native/crates/simforge-authoring/src/evaluate.rs
//! The reject filters as the authoring commands apply them: the filter mode
//! (`critical | negative-control | all`), the headline criticality band and
//! the verdict an intent rubric may combine with.

use simforge_core::evaluation::{EvaluateFilters, RejectCode, RejectFinding, Verdict};

/// `--filter`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum FilterMode {
    #[default]
    Critical,
    NegativeControl,
    All,
}

impl FilterMode {
    pub fn parse(text: &str) -> Option<Self> {
        match text {
            "critical" => Some(Self::Critical),
            "negative-control" => Some(Self::NegativeControl),
            "all" => Some(Self::All),
            _ => None,
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::Critical => "critical",
            Self::NegativeControl => "negative-control",
            Self::All => "all",
        }
    }
}

/// `filtersFor(mode, {trivialTtcS, rejectCollisions})`. `negative-control`
/// is the same filter set with `trivially_safe` demoted to a tag.
pub fn filters_for(
    mode: FilterMode,
    trivial_ttc_s: Option<f64>,
    reject_collisions: bool,
) -> EvaluateFilters {
    EvaluateFilters {
        negative_control: mode == FilterMode::NegativeControl,
        trivial_ttc_s,
        reject_collisions,
        ..EvaluateFilters::default()
    }
}

pub fn verdict_str(verdict: Verdict) -> &'static str {
    match verdict {
        Verdict::Accept => "accept",
        Verdict::Reject => "reject",
    }
}

pub fn code_str(code: RejectCode) -> String {
    match serde_json::to_value(code) {
        Ok(serde_json::Value::String(s)) => s,
        _ => String::new(),
    }
}

/// `criticalityBand(verdict, findings)`: the batch's headline bucket.
pub fn criticality_band(verdict: Verdict, findings: &[RejectFinding]) -> &'static str {
    if verdict == Verdict::Accept {
        return "critical";
    }
    let has = |code: RejectCode| findings.iter().any(|f| f.code == code);
    // Physically impossible motion is wrong before it is anything else.
    if has(RejectCode::ImplausibleMotion) {
        "implausible-motion"
    } else if has(RejectCode::NoInteraction) {
        "no-interaction"
    } else if has(RejectCode::TriviallySafe) {
        "trivially-safe"
    } else if has(RejectCode::PhysicallyUnavoidable) {
        "unavoidable"
    } else if has(RejectCode::NeverFired) {
        "never-fired"
    } else if has(RejectCode::OutOfWindow) {
        "out-of-window"
    } else {
        "trivially-safe"
    }
}

/// `combinedEvaluationVerdict`: intent evidence may explain a deliberately
/// uneventful episode, never a hard safety or execution failure.
pub fn combined_verdict(
    verdict: Verdict,
    findings: &[RejectFinding],
    intent_verdict: Option<&str>,
) -> &'static str {
    match intent_verdict {
        None => verdict_str(verdict),
        Some("reject") => "reject",
        Some(_) => {
            let explained = [
                RejectCode::NoInteraction,
                RejectCode::TriviallySafe,
                RejectCode::OutOfWindow,
            ];
            if findings.iter().any(|f| !explained.contains(&f.code)) {
                "reject"
            } else {
                "accept"
            }
        }
    }
}

/// `simforge evaluate` options.
#[derive(Debug, Clone, Default)]
pub struct EvaluateOptions {
    pub filter: FilterMode,
    pub trivial_ttc_s: Option<f64>,
    pub reject_collisions: bool,
    /// An intent rubric evaluated alongside criticality.
    pub rubric: Option<std::path::PathBuf>,
    /// Write the context-blind review packet here (needs `rubric`).
    pub blind_review_out: Option<std::path::PathBuf>,
}

/// The `filtersFor` object as the native evaluator reads it (only the keys
/// that are set).
fn filters_json(mode: FilterMode, trivial_ttc_s: Option<f64>, reject_collisions: bool) -> String {
    let mut filters = serde_json::Map::new();
    if mode == FilterMode::NegativeControl {
        filters.insert("negativeControl".into(), serde_json::Value::Bool(true));
    }
    if let Some(v) = trivial_ttc_s {
        filters.insert("trivialTtcS".into(), serde_json::json!(v));
    }
    if reject_collisions {
        filters.insert("rejectCollisions".into(), serde_json::Value::Bool(true));
    }
    serde_json::Value::Object(filters).to_string()
}

/// Keys of each criterion kind (after `id`, `kind`, `required`), as the
/// rubric schema declares them.
const CRITERION_KEYS: &[(&str, &[&str])] = &[
    ("event_order", &["mode", "interactionIds"]),
    ("trigger", &["interactionId", "outcome"]),
    ("speed_band", &["actorId", "window", "minMps", "maxMps"]),
    (
        "stationary_success",
        &["actorId", "window", "maxSpeedMps", "minPresentSeconds"],
    ),
    (
        "stop_hold_resume",
        &[
            "actorId",
            "window",
            "stopSpeedMps",
            "minHoldSeconds",
            "mustResume",
            "resumeMinSpeedMps",
            "resumeByS",
        ],
    ),
    ("clearance", &["pair", "window", "measure", "minM"]),
    ("criticality", &["metric", "pair", "window", "minS", "maxS"]),
    (
        "occlusion",
        &["observer", "target", "occluderId", "outcome"],
    ),
    ("lane_occupancy", &["actorId", "laneRsl", "mode", "window"]),
    ("zone_occupancy", &["actorId", "mode", "window", "zone"]),
    ("collision", &["pair", "maxCount"]),
    (
        "control_indication",
        &["signalId", "window", "mode", "indications"],
    ),
    ("unsupported", &["description", "reason"]),
];

fn retain(value: &mut serde_json::Value, keys: &[&str]) {
    if let serde_json::Value::Object(map) = value {
        map.retain(|k, _| keys.contains(&k.as_str()));
    }
}

/// The rubric schema's object semantics: properties it does not declare are
/// dropped, not refused (an author's note on a criterion is not an error).
fn strip_undeclared(rubric: &mut serde_json::Value) {
    retain(
        rubric,
        &["version", "intentId", "title", "originalIntent", "criteria"],
    );
    let Some(criteria) = rubric
        .get_mut("criteria")
        .and_then(serde_json::Value::as_array_mut)
    else {
        return;
    };
    for criterion in criteria {
        let kind = criterion
            .get("kind")
            .and_then(serde_json::Value::as_str)
            .map(str::to_owned);
        let Some((_, keys)) = CRITERION_KEYS
            .iter()
            .find(|(k, _)| Some(*k) == kind.as_deref())
        else {
            continue;
        };
        let mut all = vec!["id", "kind", "required"];
        all.extend_from_slice(keys);
        retain(criterion, &all);
        if let Some(zone) = criterion.get_mut("zone") {
            match zone.get("shape").and_then(serde_json::Value::as_str) {
                Some("circle") => retain(zone, &["shape", "x", "y", "radiusM"]),
                Some("box") => retain(zone, &["shape", "minX", "maxX", "minY", "maxY"]),
                _ => {}
            }
        }
    }
}

/// `readIntentRubric`: the rubric document, validated.
fn read_rubric(file: &std::path::Path) -> Result<String, simforge_compiler::CompileError> {
    use simforge_compiler::CompileError;
    let path = file.display().to_string();
    let text = std::fs::read_to_string(file).map_err(|e| {
        CompileError::at(
            "invalid_json",
            path.clone(),
            format!("cannot read intent rubric {path}: {e}"),
        )
    })?;
    let mut value: serde_json::Value = serde_json::from_str(&text).map_err(|e| {
        CompileError::at(
            "invalid_json",
            path.clone(),
            format!("cannot read intent rubric {path}: {e}"),
        )
    })?;
    strip_undeclared(&mut value);
    match simforge_core::evaluation::intent::parse_intent_rubric(&value) {
        Ok(_) => Ok(value.to_string()),
        Err(error) => {
            let issues: Vec<serde_json::Value> = match &error {
                simforge_core::error::CoreError::Schema(schema) => schema
                    .issues
                    .iter()
                    .map(|i| serde_json::json!({ "path": i.path, "reason": i.message }))
                    .collect(),
                other => vec![serde_json::json!({ "path": "", "reason": other.to_string() })],
            };
            Err(
                CompileError::at("bad_value", path, "the intent rubric is invalid")
                    .detail_entry("issues", serde_json::json!(issues))
                    .as_findings(),
            )
        }
    }
}

/// `simforge evaluate <trace>`: the reject filters (and an optional intent
/// rubric) over a trace. The document, and whether the combined verdict
/// accepts (a rejection is a finding, exit 2, not a failure to run).
pub fn run_evaluate(
    file: &std::path::Path,
    file_arg: &str,
    options: &EvaluateOptions,
) -> Result<(crate::jsvalue::JsValue, bool), simforge_compiler::CompileError> {
    use crate::jsvalue::JsValue;
    use crate::readers::{native_js, read_trace, trace_error};
    use simforge_compiler::CompileError;

    let handle = read_trace(file)?;
    let mut trace = handle.inner().clone();
    trace.quantize();
    let evaluation_text = handle
        .evaluate_json(Some(&filters_json(
            options.filter,
            options.trivial_ttc_s,
            options.reject_collisions,
        )))
        .map_err(trace_error)?;
    let evaluation: simforge_core::evaluation::TraceEvaluation =
        serde_json::from_str(&evaluation_text)
            .map_err(|e| CompileError::internal(format!("evaluation: {e}")))?;
    let evaluation_js = native_js(&evaluation_text)?;
    let band = criticality_band(evaluation.verdict, &evaluation.findings);

    let mut intent: Option<JsValue> = None;
    if let Some(rubric_file) = &options.rubric {
        let rubric = read_rubric(rubric_file)?;
        intent = Some(native_js(
            &handle.intent_rubric_json(&rubric).map_err(trace_error)?,
        )?);
        if let Some(out) = &options.blind_review_out {
            let packet = native_js(
                &handle
                    .blind_review_packet_json(&rubric)
                    .map_err(trace_error)?,
            )?;
            crate::json::write_json_file(out, &packet)?;
        }
    } else if options.blind_review_out.is_some() {
        return Err(CompileError::at(
            "missing_argument",
            "--rubric",
            "--blind-review-out requires --rubric",
        ));
    }
    let intent_verdict = intent
        .as_ref()
        .and_then(|i| i.get("verdict"))
        .and_then(JsValue::as_str)
        .map(str::to_owned);
    let combined = combined_verdict(
        evaluation.verdict,
        &evaluation.findings,
        intent_verdict.as_deref(),
    );

    let field = |name: &str| evaluation_js.get(name).cloned().unwrap_or(JsValue::Null);
    let mut entries: Vec<(String, JsValue)> = vec![
        ("file".into(), JsValue::String(file_arg.to_owned())),
        ("mapId".into(), JsValue::String(trace.header.map_id.clone())),
        (
            "metricSubject".into(),
            trace
                .header
                .metric_subject
                .clone()
                .map_or(JsValue::Null, JsValue::String),
        ),
        (
            "filter".into(),
            JsValue::String(options.filter.as_str().to_owned()),
        ),
        (
            "verdict".into(),
            JsValue::String(verdict_str(evaluation.verdict).to_owned()),
        ),
        (
            "combinedVerdict".into(),
            JsValue::String(combined.to_owned()),
        ),
        ("band".into(), JsValue::String(band.to_owned())),
        ("tags".into(), field("tags")),
        ("findings".into(), field("findings")),
        ("summary".into(), field("summary")),
        (
            "metrics".into(),
            JsValue::from_serialize(&crate::metrics::metrics_summary(&trace.metrics))?,
        ),
    ];
    if let (Some(rubric), Some(intent)) = (&options.rubric, intent) {
        entries.push((
            "intentRubric".into(),
            JsValue::String(rubric.display().to_string()),
        ));
        entries.push(("intentEvaluation".into(), intent));
    }
    if let Some(out) = &options.blind_review_out {
        entries.push((
            "blindReviewPacket".into(),
            JsValue::String(out.display().to_string()),
        ));
    }
    Ok((JsValue::object(entries), combined == "accept"))
}
